src/foundry/client-esm/applications/apps/permission-config.d.mts

```diff
@@ -1,4 +1,58 @@
+import type { USER_ROLES, UserPermission } from "../../../common/constants.d.mts";
+import type { ApplicationConfiguration, ApplicationRenderContext, ApplicationRenderOptions } from "../_types.d.mts";
 import type ApplicationV2 from "../api/application.d.mts";
 import type HandlebarsApplicationMixin from "../api/handlebars-application.mjs";
 
-export default class PermissionConfig extends HandlebarsApplicationMixin(ApplicationV2) {}
+type PermissionWithRole = UserPermission & {
+  id: string;
+  roles: {
+    name: string;
+    value: boolean;
+    readonly: string;
+  }[];
+};
+
+/**
+ * An application for configuring the permissions which are available to each User role.
+ */
+
+declare class PermissionConfig extends HandlebarsApplicationMixin(ApplicationV2) {
+  static override DEFAULT_OPTIONS: Partial<ApplicationConfiguration>;
+  static override PARTS: Record<string, HandlebarsApplicationMixin.HandlebarsTemplatePart>;
+
+  /* -------------------------------------------- */
+  /*  Rendering                                   */
+  /* -------------------------------------------- */
+
+  override _prepareContext(_options: ApplicationRenderOptions): Promise<ApplicationRenderContext>;
```

Comment:
So while the `PermissionWithRole` is currently only being used in the private method, we probably need to decide how we're handling the `_prepareContext` methods; my inclination is we should prefer proper inheritance as there's less need for people to remove unneeded defaults like we had for `getData` in AppV1, which means our return types can be stricter.

Reply:
Meaning define stricter types for the options and return value here?

Reply:
Yes, this amounts to returning `Promise<Record<string,any>>`

Reply:
I just took a cut at more explicit typing in the latest commit @JPMeehan (in permission-config.d.mts).  If that looks good, I'll do compendium-art-config.

The options parameter is unused by the method, so I removed it.  But should it be kept for consistency with the parent class?

Reply:
Definitely keep even if unused.

Reply:
Yeah if you remove it then subclasses can't access it. It also can cause some weird things.

Also if Foundry didn't name it `_options` please leave it as `options`; it shows up in intellisense which is the whole point of fvtt-types

Reply:
So call it `options`.  What should the type be (since it's not used)?  Same as the parent?

`ApplicationV2` uses a DeepPartial<RenderOptions> where AppV2 is a generic and `RenderOptions` extends `ApplicationRenderOptions`.

So should the subclass be a generic also?  Or should it define its own RenderOptions type (that would be empty)?

Reply:
- types of parameters should remain the same unless explicitly changed
- I'm personally fine with just defining it as ApplicationV2.RenderOptions. I think that interface might need a fix, the typedef has properties in it.